clients/src/main/java/org/apache/kafka/common/record/CompressionRatioEstimator.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.record;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+
+/**
+ * This class help estimate the compression ratio for each topic and compression type combination.
+ */
+public class CompressionRatioEstimator {
+    // The constant speed to increase compression ratio when a batch compresses better than expected.
+    public static final float COMPRESSION_RATIO_IMPROVING_STEP = 0.005f;
+    // The minimum speed to decrease compression ratio when a batch compresses worse than expected.
+    public static final float COMPRESSION_RATIO_DETERIORATE_STEP = 0.05f;
+    private static final ConcurrentMap<String, float[]> COMPRESSION_RATIO = new ConcurrentHashMap<>();
```

Comment:
Sorry I probably misunderstood your concern. Is it because of the new way to update compression ratio? I just ran producer-performance benchmark, the new way provides better and more stable throughput than the old one in general, which is expected as the new way should be less sensitive to the compression ratio change of a single batch.